### PR TITLE
ntrn-259: feat: query failed txs

### PR DESCRIPTION
**Pull request by NeverHappened**
implements https://p2pvalidator.atlassian.net/browse/NTRN-259

I copypasted storage extraction from oopcode request,because i cannot merge this PR (it's old)

intergrations tests for both #60 and #65 prs - https://github.com/neutron-org/neutron-integration-tests/pull/40

### CODE REVIEW

File: cmd/neutron_query_relayer/cmd/query.go
Comment (swelf19) [inline cmd/neutron_query_relayer/cmd/query.go:19]: Flag is required, i would prefer optional with a default value.

File: internal/webserver/webserver.go
Comment (swelf19) [inline internal/webserver/webserver.go:38]: Look like we need to do early return at this point

File: cmd/neutron_query_relayer/cmd/start.go
Comment (swelf19) [inline cmd/neutron_query_relayer/cmd/start.go:97]: looks like a possible race with concurrent `err` access

File: cmd/neutron_query_relayer/cmd/start.go
Comment (swelf19) [inline cmd/neutron_query_relayer/cmd/start.go:96]: what do you think about to implement graceful shutdown the web server

File: internal/app/app.go
Comment (swelf19) [inline internal/app/app.go:174]: I have a feeling, `NewDefaultStorage` should know nothing about `AllowTxQueries`.

**Review by swelf19 — COMMENTED** [review]
The following tasks can be created as inpiration from this one
1) Update intergration tests to track failed txs on relayer
2) Update documentation

File: internal/app/app.go
Comment (NeverHappened) [reply internal/app/app.go:174]: Yeah, although it's copypaste to allow e move storage to main.go, we can do something about it

Comment (NeverHappened) [reply internal/app/app.go:174]: I think it's a bit out of scope. I think current solution is good enough, but we can move it out of here later.

File: cmd/neutron_query_relayer/cmd/start.go
Comment (swelf19) [inline cmd/neutron_query_relayer/cmd/start.go:112]: defer cancelWebserverCtx()?

File: cmd/neutron_query_relayer/cmd/start.go
Comment (swelf19) [inline cmd/neutron_query_relayer/cmd/start.go:103]: call ctx's `cancel` to stop all the relayer's parts instead of logger.Fatal?

File: cmd/neutron_query_relayer/cmd/start.go
Comment (swelf19) [inline cmd/neutron_query_relayer/cmd/start.go:96]: wg.Add + defer wg.Done to wait the webserver complete shutdown gracefully

Comment (swelf19) [reply cmd/neutron_query_relayer/cmd/start.go:96]: Move all the start/stop routine into some `Run` function in the webserver pkg?

**Comment by swelf19** [comment]
I think pagination is an important feature that should be implemented

File: internal/webserver/webserver.go
Comment (swelf19) [inline internal/webserver/webserver.go:21]: unused variable

File: cmd/neutron_query_relayer/cmd/queries/utils.go
Comment (pr0n00gler) [inline cmd/neutron_query_relayer/cmd/queries/utils.go:15]: Not so important (maybe it's even an overhead here, so I'm not requesting changes), but I think it's a good practice to use url builder instead of just string concatenation. It helps to avoid silly mistakes like double slash, etc.

File: internal/webserver/webserver.go
Comment (NeverHappened) [reply internal/webserver/webserver.go:20]: Right now some of the contexts are defined (monitoring and main) in their own packages, that's why I'm gonna stick to it

File: internal/storage/leveldb.go
Comment (oopcode) [inline internal/storage/leveldb.go:135]: please wrap the error

File: internal/storage/leveldb.go
Comment (oopcode) [inline internal/storage/leveldb.go:155]: please wrap the error

File: internal/storage/leveldb.go
Comment (oopcode) [inline internal/storage/leveldb.go:140]: please wrap the error

File: internal/webserver/webserver.go
Comment (oopcode) [inline internal/webserver/webserver.go:67]: this functionality can 100% be covered with integration tests; please add them

**Review by oopcode — CHANGES_REQUESTED** [review]

File: internal/webserver/webserver.go
Comment (sotnikov-s) [inline internal/webserver/webserver.go:49]: do we really need this complicated overloading shutting down for such a simple webserver? it won't hurt if we remove it, but it would simplify code here and in the start.go to a simple `go webserver.Run` call. we barely care about opened and idle connections that are gracefully closed in the Shutdown method, it's a kind of internal usage readonly service

**Review by sotnikov-s — CHANGES_REQUESTED** [review]

File: cmd/neutron_query_relayer/cmd/root.go
Comment (NeverHappened) [reply cmd/neutron_query_relayer/cmd/root.go:10]: I think I used it in other package. If not anymore, can make it private again

File: internal/webserver/webserver.go
Comment (NeverHappened) [reply internal/webserver/webserver.go:49]: I think we already have it, so why delete it.
Also:
Webserver will also handle more complicated actions, such as resubmit failed tx (see [task](https://github.com/neutron-org/neutron-query-relayer/pull/65/files))
This action would probably require more careful shutdown

File: cmd/neutron_query_relayer/cmd/start.go
Comment (NeverHappened) [reply cmd/neutron_query_relayer/cmd/start.go]: I see one point for it staying as it were:
It's just hard separates these two servers so they are independent.
Why:
- We can configure metrics and webserver access in different way (because different ports)
- Some kind of trouble in one thing does not bring down another (debatable, but still)

Comment (sotnikov-s) [reply cmd/neutron_query_relayer/cmd/start.go]: > We can configure metrics and webserver access in different way (because different ports)

we can, but why? isn't different endpoints enough?

> Some kind of trouble in one thing does not bring down another (debatable, but still)

what can go wrong with a web server? it just handles requests. if a request goes bad, only this request's scope is affected, the web server execution persists

Comment (NeverHappened) [reply cmd/neutron_query_relayer/cmd/start.go]: agreed, other things can be resolved with reverse-proxy

File: internal/webserver/webserver.go
Comment (sotnikov-s) [inline internal/webserver/webserver.go:25]: there is a HandlerFunc already in the http package, so this can be removed

File: internal/webserver/webserver.go
Comment (sotnikov-s) [inline internal/webserver/webserver.go:70]: this func should be private

File: internal/webserver/monitoring.go
Comment (sotnikov-s) [inline internal/webserver/monitoring.go:30]: this func should be private

File: internal/webserver/webserver.go
Comment (sotnikov-s) [reply internal/webserver/webserver.go:70]: actually there is a lot of public stuff in the package that is not used from the outside. please reconsider publicity in the package

File: internal/webserver/monitoring.go
Comment (sotnikov-s) [inline internal/webserver/monitoring.go:39]: what's the meaning of this call in this method?

Comment (swelf19) [reply internal/webserver/monitoring.go:39]: I can see two ways to achieve the goal to get amount of unsuccessful txs:
1) eval ` SetUnsuccessfulTxsSizeQueue` on every `saveIntoErrorQueue` and `removeFromErrorQueue`, but i do not feel it right to inject the global metrics into the storage, may be i'm wrong, we already have the metrics spread throughout the code
2) measure the amount of txs during the `/metrics` request

Comment (sotnikov-s) [reply internal/webserver/monitoring.go:39]: re 2:
you now eval `SetUnsuccessfulTxsSizeQueue` twice on each request: first in `ServeHTTP` since it's called on each incoming request, and second in the `UnsuccessfulTxs` since it's the `/metrics` endpoint handler. so, at least you can delete the `FillUnsuccessfulTxsMetric` call and the func itself and keep calculating only in the `metrics` handler.

re 1:
I find this solution as the best option here. I'm totally fine with importing metric setters across the app, I believe it's the way they are supposed to be used

Comment (swelf19) [reply internal/webserver/monitoring.go:39]: i've changed a mind about first approach.
you need an extra storage-metric specific code to determine is it a new key-value or a modification existing one https://github.com/neutron-org/neutron-query-relayer/blob/feat/query-failed-txs/internal/storage/leveldb.go#L231. I'm against extra code in the storage only for metrics (`Has` method)

File: internal/webserver/webserver.go
Comment (swelf19) [reply internal/webserver/webserver.go]: Just the opposite.
error `http.ErrServerClosed` is initialized by server.Shutdown few lines below, which is initialized by canceling context, we need do nothing more
error != http.ErrServerClosed something goes wrong, we need to initialise the app shutdown.
But now i got your point about shutdown routine comlexity

File: internal/webserver/monitoring.go
Comment (sotnikov-s) [reply internal/webserver/monitoring.go]: re 2: sorry, I'm wrong, ignore
re 1: okay, I see, it's not that easy indeed

I'm fine with the current version, don't think it's gonna hurt or slow the app down in any way

**Review by sotnikov-s — COMMENTED** [review]
LGTM. waiting for integration tests before approval

File: .env.example
Comment (foxpy) [inline .env.example:33]: I believe it should be possible to define not just a port, but a whole URL to bind to, e.g. `127.0.0.1:9999`

File: .env.example.dev
Comment (foxpy) [inline .env.example.dev:25]:
```suggestion
RELAYER_WEBSERVER_PORT=127.0.0.1:9999
```

File: .env.example
Comment (foxpy) [inline .env.example:33]:
```suggestion
RELAYER_WEBSERVER_PORT=127.0.0.1:9999
```

**Review by foxpy — CHANGES_REQUESTED** [review]

**Review by foxpy — DISMISSED** [review]

**Review by sotnikov-s — DISMISSED** [review]
tACK

File: internal/storage/leveldb.go
Comment (pr0n00gler) [inline internal/storage/leveldb.go:196]: `removeFromUnsuccessfulQueue` doesn't return an error if tx is not in the UnsuccessfulQueue, right?
So, there are no problems with this `removeFromUnsuccessfulQueue` call, if a tx was not saved into `UnsuccessfulQueue` and immediately goes with `relay.Committed` status here?

**Review by foxpy — APPROVED** [review]

**Review by sotnikov-s — APPROVED** [review]

**Review by pr0n00gler — APPROVED** [review]

**Review by oopcode — APPROVED** [review]

**Pull request merged**
